libraries/GCS_MAVLink/MissionItemProtocol_Waypoints.cpp

```diff
@@ -147,4 +151,128 @@ void MissionItemProtocol_Waypoints::truncate(const mavlink_mission_count_t &pack
     mission.truncate(packet.count);
 }
 
+// returns a unique ID for this mission
+bool MissionItemProtocol_Waypoints::opaque_id(uint32_t &checksum) const
+{
+    WITH_SEMAPHORE(mission.get_semaphore());
+    switch (checksum_state.state) {
+    case ChecksumState::READY:
+        if (mission.last_change_time_ms() != checksum_state.mission_change_time_ms) {
+            return false;
+        }
+        checksum = checksum_state.checksum;
+        // can't use zero as the field is an extension field in mavlink2:
+        if (checksum == 0) {
+            checksum = UINT32_MAX;
+        }
+        return true;
+    case ChecksumState::CALCULATING:
+    case ChecksumState::ERROR:
+        return false;
+    }
+    return false;
+}
+
+void MissionItemProtocol_Waypoints::invalidate_checksum()
+{
+    WITH_SEMAPHORE(mission.get_semaphore());
+
+    checksum_state.state = ChecksumState::CALCULATING;
+    checksum_state.checksum = 0;
+    checksum_state.current_waypoint = 1;
+    checksum_state.count = mission.num_commands();
+    checksum_state.mission_change_time_ms = mission.last_change_time_ms();
+}
+
+void MissionItemProtocol_Waypoints::update_checksum()
+{
+    // update the checksum if required:
+
+    WITH_SEMAPHORE(mission.get_semaphore());
+
+    const uint32_t mission_last_change_time_ms = mission.last_change_time_ms();
+
+    if (mission_last_change_time_ms == checksum_state.last_calculate_time_ms) {
```

Comment:
there could be more than one change in a ms, especially with scripting changing missions, or DDS